// typedb-client/connection/cluster/ClusterServerClient.js
"use strict";
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClusterServerClient = void 0;
const TypeDBClientImpl_1 = require("../TypeDBClientImpl");
const ClusterServerStub_1 = require("./ClusterServerStub");
const TypeDBDatabaseManagerImpl_1 = require("../TypeDBDatabaseManagerImpl");
const TypeDBClientError_1 = require("../../common/errors/TypeDBClientError");
const ErrorMessage_1 = require("../../common/errors/ErrorMessage");
var CLIENT_NOT_OPEN = ErrorMessage_1.ErrorMessage.Client.CLIENT_NOT_OPEN;
class ClusterServerClient extends TypeDBClientImpl_1.TypeDBClientImpl {
    constructor(address, credential) {
        super();
        this._address = address;
        this._credential = credential;
        this._isOpen = false;
    }
    async open() {
        await super.open();
        this._stub = new ClusterServerStub_1.ClusterServerStub(this._address, this._credential);
        await this._stub.open();
        this._databases = new TypeDBDatabaseManagerImpl_1.TypeDBDatabaseManagerImpl(this._stub);
        this._isOpen = true;
        return this;
    }
    isOpen() {
        return this._isOpen;
    }
    stub() {
        return this._stub;
    }
    get databases() {
        if (!this.isOpen())
            throw new TypeDBClientError_1.TypeDBClientError(CLIENT_NOT_OPEN);
        return this._databases;
    }
    async close() {
        if (this.isOpen()) {
            this._isOpen = false;
            await super.close();
            this._stub.close();
        }
    }
}
exports.ClusterServerClient = ClusterServerClient;
